### ISSUE TITLE
Feature Request: Support DI for migrations by allowing non-empty constructors

### ISSUE THREAD

**Issue by nphmuller** [aspnet/EntityFrameworkCore#10660]
Currently (as of version 2.0) when a migration is created with a non-empty constructor the following exception is thrown during migration:
```
System.MissingMethodException: No parameterless constructor defined for this object.
````

Since ConfigureServices is already called during migrations and the ServiceProvider is used to resolve the DbContext it would be really nice if the same ServiceProvider could be used to resolve the constructor of migrations.

This would make it possible to inject other services and run them during the migration.

My use-case:
We use ElasticSearch as a search index on top of some of our ef model types. When the model scheme is changed, the search index scheme also has to be updated and sometimes data has to be reindexed.

Previously we simply used to Migration Id to check if the search index is at the same scheme as the database. If the search index migration id was not the same as the one in the database we simply recreated the index and reindexed all the data.

Recently our data grow to a size that this is no longer feasonable. It would take way too long. So we only want to update the search index during migrations that would affect the parts of the scheme that are indexed.

So what I wanted to do is call the (for example) reindex method from an ef migration, when the search index should be updated. For this I would need the inject the ElasticSearch client into the migration, which is not supported atm. Hence the feature request.

**Comment by nphmuller** [comment]
Seems this worked (although unsupported) in 1.0-RC1: https://github.com/aspnet/EntityFrameworkCore/issues/6574

**Comment by nphmuller** [comment]
I also tried making a custom operation by overriding the default Sql generator with my own custom one, but when I try to inject the dependency (elastic search client) I get the following error:
```
Unable to resolve service for type 'Nest.IElasticClient' while attempting to activate 'MyApp.Migrations.MyMigrationsSqlGenerator'.
```

**Comment by nphmuller** [comment]
Onto the next problem. The error in my previous comment means I have to register IElasticClient as a DesignTimeService. I can do this by implementing IDesignTimeServices and configuring the service in that class.

The problem is that ElasticClient takes some configuration settings from appsettings.json. However to get to proper appsettings I need to know the hosting environment. Currently I can't find a way to obtain the proper IHostingEnvironment from my IDesignTimeServices implementation.

**Comment by ralmsdeveloper** [comment]
Is this what you need?
https://github.com/aspnet/EntityFrameworkCore/issues/9415#issuecomment-347944628

**Comment by nphmuller** [comment]
@ralmsdeveloper Nope. I'm not trying to create the DbContext at design-time. That already works perfectly. I'm trying to create inject one of my own services in a Migration or a custom SqlServerMigrationsSqlGenerator. See [this](https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/operations#using-a-migrationoperation).

**Comment by nphmuller** [comment]
I first tried to add IElasticClient to the design time services by implementing IDesignTimeServices. Turns out the internal service provider does not have access to the customized design time services, so I had to swap out the internal service provider for the main service provider. 

Basically I had to turn this:
``` c#
services.AddDbContext<MyContext>(opt =>
{
    opt.UseSqlServer(configuration.GetConnectionString("DefaultConnection"))
       .ReplaceService<IMigrationsSqlGenerator, MyMigrationsSqlGenerator>();
});
```` 
Into this:
``` c#
services.AddEntityFrameworkSqlServer();
services.AddSingleton<IMigrationsSqlGenerator, MyMigrationsSqlGenerator>();

services.AddDbContext<MyContext>((sp, opt) =>
{
    opt.UseSqlServer(configuration.GetConnectionString("DefaultConnection"))
       .UseInternalServiceProvider(sp);
});
```

**Comment by ajcvickers** [comment]
@nphmuller While we can see the value in this when used correctly, we have been a little reluctant to implement it because it can also become a pit of failure. Specifically, migrations are used to generate the SQL used to update the database, but it is not executed at that time. Indeed, if scripting is being used it may not even be executed at all on the machine that is generating. Typically when people have asked for this before it has been to "connect to the database" to manipulate it in some way while migrations are being generated. This we want to avoid. Leaving untriaged now to discuss with others when they are back.

**Comment by nphmuller** [comment]
@ajcvickers I can see that. Didn't really think about the script generation feature, since we don't use it, but that is a good point. 

What's documented [here](https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/operations#using-a-migrationoperation) provided a good base for me to implement my use-case anyway. Since `IMigrationsSqlGenerator` supports DI. The only thing that was really confusing for me is that I should replace the internal service provider, instead of providing an implementation of `IDesignTimeServies`, since I though `IMigrationsSqlGenerator` was part of the design time service. But I had no previous experience with this part of EF, so that didn't help.

**Comment by ajcvickers** [comment]
We discussed this in triage and decided that we could make the service provider available to migrations via IInfrastructure. This unblocks people from doing advanced things while not promoting too much of a pit of failure.
